tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/DumpFilesModelCheckerListener.scala

```diff
@@ -20,28 +20,30 @@ import com.typesafe.scalalogging.LazyLogging
  */
 object DumpFilesModelCheckerListener extends ModelCheckerListener with LazyLogging {
 
-  override def onCounterexample(
-      rootModule: TlaModule,
-      trace: DecodedExecution,
-      invViolated: TlaEx,
-      errorIndex: Int): Unit = {
-    dump(rootModule, trace, invViolated, errorIndex, "violation")
+  override def onCounterexample(counterexample: Counterexample, errorIndex: Int): Unit = {
+    dump(counterexample, errorIndex, "violation")
   }
 
   override def onExample(rootModule: TlaModule, trace: DecodedExecution, exampleIndex: Int): Unit = {
-    dump(rootModule, trace, tla.bool(true).as(BoolT1), exampleIndex, "example")
+    val counterexample = Counterexample(rootModule, trace, tla.bool(true).as(BoolT1))
+    dump(counterexample, exampleIndex, "example")
   }
 
   private def dump(
-      rootModule: TlaModule,
-      trace: DecodedExecution,
-      invViolated: TlaEx,
+      counterexample: Counterexample,
       index: Int,
       prefix: String): Unit = {
-    val states = trace.path.map(p => (p._2.toString, p._1))
-
     def dump(suffix: String): List[String] = {
-      CounterexampleWriter.writeAllFormats(prefix, suffix, rootModule, invViolated, states)
+      // TODO(shonfeder): Should the CounterexampleWriter take a Counterexample?
+      // Would require fixing inter-package dependencies, since it would require
+      // exposing the Counterexample class to the tla-io project.
```

Comment:
It looks to me like `Counterexample` is very much tuned towards the purposes of `tla-bmcmt`. So it should probably belong there.

Reply:
I guess I can do that, if you think the organization makes sense. But I don't know where the associated methods in the companion object would belong. Also, as noted above, this `states` structure seems superfluous, and we should really be using a `DecodedExecution` imo. But then that is just a wrapper around an association list too.... I think I'm not sure how to impose a more sensible organization without a deeper refactor, lol. Any specific tips for incremental improvement on this PR welcome tho!

Reply:
I was able to work this out by just dropping reference to the `DecodeExecution`.